File: problem-solving/_lc-62-medium-unique-paths/3-bottom-up-space-optimized.py
"""
https://leetcode.com/problems/unique-paths/

Complexity
----------
Time: O(m*n)
Space: O(n)
"""

class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        curr = [0] * (n + 1)
        prev = curr[:]
        for r in range(1, m + 1):
            for c in range(1, n + 1):
                if r == 1 or c == 1:
                    curr[c] = 1
                    continue
                curr[c] = prev[c] + curr[c - 1]
            prev = curr[:]
        return curr[n]

File: problem-solving/_lc-1220-hard-count-vowels-permutation/3-bottom-up-space-optimized.go
/*
https://leetcode.com/problems/count-vowels-permutation/

Complexity
----------
Time: O(n)
Space: O(1)
*/

func countVowelPermutation(n int) int {
    M := 1000000007
    a, e, i, o, u := 1, 1, 1, 1, 1
    for idx := 1; idx < n; idx++ {
        // Previous counts
        pa, pe, pi, po, pu := a, e, i, o, u

        // a can be after e, i or u, and so on...
        a = (pe + pi + pu) % M
        e = (pa + pi) % M
        i = (pe + po) % M
        o = pi
        u = (pi + po) % M
    }
    return (a + e + i + o + u) % M
}
